airflow/models/taskinstance.py

```diff
@@ -553,6 +553,26 @@ def clear_task_instances(
     session.flush()
 
 
+def record_task_instance_history(ti, session):
+    """Update the task instance history table."""
+    if ti.state not in [
```

Comment:
this seems a bit odd

i would think that we would know, when we call it, whether it should be recorded

Reply:
Yeah. I was avoiding using session.merge and rewriting history. I changed this to a query to check if the object exist before inserting. That way, we won't have integrity error because repeatedly clearing some states would cause integrity error since the try_number won't have changed